meetings/working-groups/collection-literals/Supporting-Interface-TargetTypes.md

```diff
@@ -0,0 +1,72 @@
+
+   As before, promising too much can hurt us (and defy user expectations around performance and safety).
+
+5. Certain concrete types (specifically `List<T>` or `T[]`) have highly undesirable downsides for an interface target.  Specifically, it would make it a very dangerous footgun for any API to expose an `IEnumerable<T>/IReadOnlyCollection<T>/IReadOnlyList<T>` property that was instantiated with a collection expression.  A set of users would be very unhappy and perturbed to discover that they switched away from an existing read-collection type to a collection expression and now had things become unsafe.  This would tie into the feeling that collection expressions could not be trusted to make sound decisions that people could trust.
+
+## Options
+
+If we picked an existing concrete collection type to instantiate when targeting an interface type, the choices would likely be the following:
+
+1. `List<T>`.  The most flexible.  But also not very efficient, and definitely unsafe to expose through an `IEnumerable<T>/IReadOnlyCollection<T>/IReadOnlyList<T>` property.  
+2. `T[]`.  A strange middle ground.  Only partially flexibly.  Can mutate the values in the array themselves, but cannot do anything affecting the length of the array.  Also unsafe to expose through an `IEnumerable<T>/IReadOnlyCollection<T>/IReadOnlyList<T>` property.  
+3. `ImmutableArray<T>`.  The least flexible.  Provides safety for elements.  But now does not work well for the case of passing to `ICollection<T>/IList<T>` (which can mutate the number of elements).
+
+If we don't pick a concrete type, we have a few options at our disposal:
+
+1. State you only get an unnamed impl of *only* that target interface, and nothing else.  For example, if you target typed to `IEnumerable<T>`, you would only get a type type implemented `IEnumerable<T>` and nothing else.  This would be entirely safe (no worry about something mutating your values, or taking a dependency on a concrete type).  However, it would also leave perf on the table.  Many BCL apis optimize themselves by querying for `IList<T>` to take a fast path if the value exposes that interface.
+
+2. State you get an unnamed impl of a type that is guaranteed to implement both `IReadOnlyList<T>` and `IList<T>` *and* that that type returns `true` for `ICollection<T>.IsReadOnly` or not, depending on what you were targeting.  If you were targeting only an `IEnumerable<T>/IReadOnlyCollection<T>/IReadOnlyList<T>` you would get something read-only.  If you asked for an `IList<T>/ICollection<T>` it would be non-read-only.  This would have the benefit of being entirely safe for the non-mutation cases (because the collection IsReadOnly, consumers could not mutate elements).  It would also be something highly optimizable.  BCL methods would query for `IList<T>` and do fast paths for it.  We could also still fast-path our `GetEnumerator`` impl to be low allocs.
```

Comment:
```suggestion
2. State you get an unnamed impl of a type that is guaranteed to implement both `IReadOnlyList<T>` and `IList<T>` *and* that that type returns `true` for `ICollection<T>.IsReadOnly` or not, depending on what you were targeting.  If you were targeting only an `IEnumerable<T>/IReadOnlyCollection<T>/IReadOnlyList<T>` you would get something read-only.  If you asked for an `ICollection<T>/IList<T>` it would be non-read-only.  This would have the benefit of being entirely safe for the non-mutation cases (because the collection IsReadOnly, consumers could not mutate elements).  It would also be something highly optimizable.  BCL methods would query for `IList<T>` and do fast paths for it.  We could also still fast-path our `GetEnumerator`` impl to be low allocs.
```